rlpr-2.06/man/rlprrc.5
.\" SCCS "@(#)rlprrc.5	1.3	01/01/02 meem"
.TH rlprrc 5 "1999/10/28" "rlpr 2.04" "UNIX Reference Manual"
.SH NAME
rlprrc \- remote printing resource file
.SH DESCRIPTION
The \fBrlprrc\fP configuration file is consulted by the remote
printing commands (\fBrlpr\fP(1), \fBrlpq\fP(1), and \fBrlprm\fP(1))
to resolve a printqueue or a hostname. It can be used to either lookup
a printqueue (printer) for a given host, or a host for a given
printqueue.  If both the printer and the printhost are known, this
file is never consulted.
.PP
Initially, the remote printing commands look for the file \fI.rlprrc\fR
in the user's home directory.  If that file cannot be found or does
not provide the information necessary to resolve the query, then
the system-wide \fI/etc/rlprrc\fR is consulted (if present).
If resolution is still unsuccessful, the command fails.
.PP
There is one entry per line \- each line has the format:
.PP
\fBhostname: printer1 printer2 ... printerN\fP
.PP
Which indicates that host \fBhostname\fP has printers \fBprinter1
printer2 ... printerN\fP available.  In the case where the remote
printing command knows the host to print to but does not know which
printer to use on that host, the first one (\fBprinter1\fP here) will
be used.
.PP
In the case where the remote printing command knows the printqueue
(printer) to print to but does not know what host has that queue, the
last host which has the specified printqueue is used.  If this
behavior is not acceptable, you can force a certain host to always be
used for a certain printqueue by putting a \fB!\fP after the
printqueue name.
.PP
For example, an \fB.rlprrc\fP file containing:
.ta \w'1234567'u
.nf

	foo.baz.org: litho laserjet4! 
	foo.bar.org: laserjet4 lineprinter

.fi
.PP
will always resolve the printqueue \fBlaserjet4\fP to
\fBfoo.baz.org\fP no matter what other hosts have a printqueue named
\fBlaserjet4\fP.
.PP
The \fBrlprrc\fP file is only consulted in situations where a
printqueue or hostname needs to be resolved.  \fBIt is provided only for
convenience and is not required.\fP
.SH SEE ALSO
\fBrlpr\fP(1), \fBrlpq\fP(1), \fBrlprm\fP(1), \fBrlprd\fP(8)
.SH AUTHOR
meem <meem@gnu.org>


rlpr-2.06/NEWS
Changes in release 2.06:

* fixed several autoconf-related problems introduced in 2.05.
* support fallback to syslog() when vsyslog() is not available.

Changes in release 2.05:

* fixed security flaw detailed at: http://www.osvdb.org/7194

Changes in release 2.04:

* removed -W flag so that it does not conflict with POSIX.2 systems.
* fixed rlpr to not expect an argument with --windows (nee -W) option.
* fixed `bsdlinks' target to not accidentally remove lprm.
* stubbed out i18n/l10n functions to make ./configure --disable-nls work.
* enhanced rlpr to print a warning if the user attempts to print an
  empty file.
  
Changes in release 2.03:

* added support for printer alternation in check-server.sh.
* fixed message displayed when incorrect non-long style options are
  passed to rlpr, rlpq, rlprd, or rlprm.
* added configurable configuration directory (RLPR_CONFDIR) support.
* added support for running `configure' out of other than `top_srcdir'
* fixed build problem that caused libnsl and libsocket (when required)
  not to appear on link line.
* new autoconf macros for EXIT_FAILURE/EXIT_SUCCESS.
* fixed two bugs in rlprrc lookups of hosts for a printer.
* fixed bug which prevented rlprrc lookups for a printer for a host
  (now, default to `lp' is only done when the rlprrc lookup fails).

Changes in release 2.02:

* fixed connect() to report the correct error message upon failure.
* changed Makefiles to never use += operator (for portability).
* revised code to compile correctly on newer releases of GNU libc.
* updated pointers to packages in README.
* fixed --no-bind to work when running as root.
* fixed Makefiles so error.o is now only put in librlpr once.
* fixed broken configure.in custom checks for MAXHOSTNAMELEN and off_t.
* added custom check for h_errno definition in <netdb.h>.
* fixed test(1) brokenness in check-server.sh.

Changes in release 2.01:

* fixed bug that would cause rlpr to send from unprivileged ports
* fixed some longstanding argument parsing problems
* added `--hostname' option
* fixed bug that caused rlpr to be interrupted with SIGPOLL
* added generic support for vendor extensions to lpr
* added support for some hp/ux lpr extensions
* changed the default timeout to be more reasonable
* small fixes to make the code compile properly on aix 4.1.x
* small fixes to make the code compile properly on freebsd
* made ~/.rlprrc truly optional

Changes in release 2.00:

* complete rewrite of rlpr, rlpq, and rlprd - much more modular
* addition of new rlprm command
* complete re-architecture of build environment (deep directories now)
* new hooks for internationalization support
* build procedure now uses automake in addition to autoconf
* remote printing commands now support a --timeout option in case
  remote end dies
* new --send-data-first option to rlpr to work with buggy lpd's
* rlpr now works with hp jetdirect printers
* rlpr now includes a shiny-new progress meter, useful when printing
  files over slow links
* all printing commands default to `lp' printer now, like bsd
* RLPR_PROXY has been removed; use RLPR_PROXYHOST instead
* uses automake, autoconf, etc..


rlpr-2.06/src/rlprrc.c
/*
 * Copyright (c) 1998-1999 peter memishian (meem), meem@gnu.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

#pragma ident "@(#)rlprrc.c	1.3	99/10/29 meem"

#include <string.h>
#include <stdio.h>
#include <lib.h>
#include <errno.h>
#include <stdlib.h>		/* getenv() */

#include "msg.h"
#include "rlprrc.h"
#include "intl.h"

static const char *
rlprrc_search_internal(FILE *rcfile, const char *key, rlpr_search_type_e type)
{
    char	line_buf[R_LINE_LEN];
    char       *queue_pos, *ptr, *line;
    char       *result = 0;
    size_t	keylen = strlen(key);

    msg(R_DEBUG, 0, "rlprrc_search_internal: resolving `%s' (type %d)", key,
	type);

    while (fgets(line_buf, R_LINE_LEN, rcfile) != 0) {

	line = line_buf;
	while (strchr(" \t", *line) != 0)
	    line++;

	/* skip comments or invalid lines */
	if (strchr("#\n", *line) || (queue_pos = strchr(line, ':')) == 0)
	    continue;

	while (strchr(": \t", *queue_pos) != 0)
	    *queue_pos++ = '\0';

	if (type == R_RESOLVE_HOST) {

	    while (strncmp(queue_pos, key, keylen) != 0 ||
		   strchr(" !\n", queue_pos[keylen]) == 0) {
		queue_pos = strpbrk(queue_pos, " \t");
		if (queue_pos == 0)
		    break;
		while (strchr(" \t", *queue_pos) != 0)
		    *queue_pos++ = '\0';
	    }

	    if (queue_pos == 0)
		continue;

	    free(result);
	    result = xstrdup(line);
	    if (queue_pos[keylen] == '!')
		break;

	} else if (type == R_RESOLVE_QUEUE) {

	    if (strcasecmp(line, key) != 0)
		continue;

	    if ((ptr = strpbrk(queue_pos, "\n ")) != 0)
		*ptr = '\0';

	    return xstrdup(queue_pos);
	}
    }

    return result;
}

const char *
rlprrc(const char *key, rlpr_search_type_e type)
{
    FILE       *rcfile = 0;
    char       *home = getenv("HOME");
    char       *confdir = getenv("RLPR_CONFDIR");
    char       *rlprrc_name;
    const char *result = 0;

    if (home != 0) {

	rlprrc_name = xmalloc(strlen(home) + strlen(R_RLPRRC_NAME) + 3);
	sprintf(rlprrc_name, "%s/.%s", home, R_RLPRRC_NAME);

	rcfile = fopen(rlprrc_name, "r");
	free(rlprrc_name);

	if (rcfile != 0)
	    result = rlprrc_search_internal(rcfile, key, type);
	else
	    msg(R_DEBUG, errno, "unable to open ~/.%s", R_RLPRRC_NAME);
    } else
	msg(R_WARNING, 0, "$HOME not set, not using ~/.%s", R_RLPRRC_NAME);

    if (result == 0) {

	if (confdir == 0)
	    confdir = R_CONFDIR;

	rlprrc_name = xmalloc(strlen(confdir) + strlen(R_RLPRRC_NAME) + 3);
	sprintf(rlprrc_name, "%s/%s", confdir, R_RLPRRC_NAME);

	msg(R_DEBUG, 0, "`%s' not found in ~/.%s, checking %s", key,
	    R_RLPRRC_NAME, rlprrc_name);

	rcfile = fopen(rlprrc_name, "r");
	if (rcfile != 0)
	    result = rlprrc_search_internal(rcfile, key, type);
	else
	    msg(R_DEBUG, errno, "unable to open %s", rlprrc_name);

	free(rlprrc_name);
    }

    if (rcfile && fclose(rcfile) == -1)
	msg(R_WARNING, errno, "unable to close %s", R_RLPRRC_NAME);

    msg(R_DEBUG, 0, "rlprrc: `%s' -> `%s' (type %d)", key, result ? result :
	_("<not resolved>"), type);
    return result;
}
